File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/rules/naming.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Callable, Optional

from ..types import Issue, Parsed


@dataclass
class _Rule:
    code: str
    severity: str
    field: str
    message: str
    check: Callable


NAMING_RULES: list[_Rule] = [
    _Rule(
        "E000_NAME_MISSING",
        "error",
        "name",
        "name is required",
        lambda p, d, o: not p.name,
    ),
    _Rule(
        "E001_NAME_TOO_LONG",
        "error",
        "name",
        "name must be 1-64 chars",
        lambda p, d, o: bool(p.name) and len(p.name) > 64,
    ),
    _Rule(
        "E002_NAME_BAD_CHAR",
        "error",
        "name",
        "name must use only [a-z0-9-]",
        lambda p, d, o: bool(p.name) and not re.fullmatch(r"[a-z0-9-]+", p.name),
    ),
    _Rule(
        "E003_NAME_HYPHEN_EDGE",
        "error",
        "name",
        "name cannot start or end with '-'",
        lambda p, d, o: bool(p.name)
        and (p.name.startswith("-") or p.name.endswith("-")),
    ),
    _Rule(
        "E004_NAME_CONSECUTIVE_HYPHENS",
        "error",
        "name",
        "name cannot contain '--'",
        lambda p, d, o: bool(p.name) and "--" in p.name,
    ),
    _Rule(
        "E005_NAME_DIRECTORY_MISMATCH",
        "error",
        "name",
        "name must equal parent directory name",
        lambda p, d, o: bool(p.name) and bool(d) and p.name != d,
    ),
    _Rule(
        "E006_TIER_MISSING",
        "error",
        "tier",
        "tier is required",
        lambda p, d, o: not p.tier,
    ),
    _Rule(
        "E007_TIER_INVALID",
        "error",
        "tier",
        "tier must be one of org|team|app|project",
        lambda p, d, o: bool(p.tier) and p.tier not in o["allowed_tiers"],
    ),
    _Rule(
        "E008_DESCRIPTION_MISSING",
        "error",
        "description",
        "description is required",
        lambda p, d, o: not p.description,
    ),
    _Rule(
        "E012_DESCRIPTION_TOO_LONG",
        "error",
        "description",
        "description must be 1-1024 chars",
        lambda p, d, o: bool(p.description) and len(p.description) > 1024,
    ),
    _Rule(
        "E009_DESCRIPTION_TOO_SHORT",
        "warn",
        "description",
        "description < 60 chars; may not trigger reliably",
        lambda p, d, o: bool(p.description) and len(p.description) < 60,
    ),
]

_DEFAULT_TIERS = ["org", "team", "app", "project"]


def check_naming(
    parsed: Parsed,
    directory: Optional[str],
    opts: Optional[dict] = None,
) -> list[Issue]:
    opts = opts or {}
    extra = opts.get("extra_tiers", [])
    allowed = opts.get("allowed_tiers") or _DEFAULT_TIERS + list(extra)
    merged = {"allowed_tiers": allowed}
    issues: list[Issue] = []
    for r in NAMING_RULES:
        if r.check(parsed, directory, merged):
            issues.append(Issue(r.code, r.severity, r.field, r.message))
    return issues

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/rules/naming.mjs
import { makeIssue } from "../types.mjs";

export const NAMING_RULES = [
  {
    code: "E000_NAME_MISSING",
    severity: "error",
    field: "name",
    message: "name is required",
    check: (p) => !p.name,
  },
  {
    code: "E001_NAME_TOO_LONG",
    severity: "error",
    field: "name",
    message: "name must be 1-64 chars",
    check: (p) => !!p.name && p.name.length > 64,
  },
  {
    code: "E002_NAME_BAD_CHAR",
    severity: "error",
    field: "name",
    message: "name must use only [a-z0-9-]",
    check: (p) => !!p.name && !/^[a-z0-9-]+$/.test(p.name),
  },
  {
    code: "E003_NAME_HYPHEN_EDGE",
    severity: "error",
    field: "name",
    message: "name cannot start or end with '-'",
    check: (p) => !!p.name && (/^-/.test(p.name) || /-$/.test(p.name)),
  },
  {
    code: "E004_NAME_CONSECUTIVE_HYPHENS",
    severity: "error",
    field: "name",
    message: "name cannot contain '--'",
    check: (p) => !!p.name && /--/.test(p.name),
  },
  {
    code: "E005_NAME_DIRECTORY_MISMATCH",
    severity: "error",
    field: "name",
    message: "name must equal parent directory name",
    check: (p, dir) => !!p.name && !!dir && p.name !== dir,
  },
  {
    code: "E006_TIER_MISSING",
    severity: "error",
    field: "tier",
    message: "tier is required",
    check: (p) => !p.tier,
  },
  {
    code: "E007_TIER_INVALID",
    severity: "error",
    field: "tier",
    message: "tier must be one of org|team|app|project",
    check: (p, _dir, opts) =>
      !!p.tier && !opts.allowedTiers.includes(p.tier),
  },
  {
    code: "E008_DESCRIPTION_MISSING",
    severity: "error",
    field: "description",
    message: "description is required",
    check: (p) => !p.description,
  },
  {
    code: "E012_DESCRIPTION_TOO_LONG",
    severity: "error",
    field: "description",
    message: "description must be 1-1024 chars",
    check: (p) => !!p.description && p.description.length > 1024,
  },
  {
    code: "E009_DESCRIPTION_TOO_SHORT",
    severity: "warn",
    field: "description",
    message: "description < 60 chars; may not trigger reliably",
    check: (p) => !!p.description && p.description.length < 60,
  },
];

const DEFAULT_TIERS = ["org", "team", "app", "project"];

export function checkNaming(parsed, dirName, opts = {}) {
  const extraTiers = opts.extraTiers ?? [];
  const allowedTiers = opts.allowedTiers ?? [...DEFAULT_TIERS, ...extraTiers];
  const merged = { allowedTiers };
  const issues = [];
  for (const r of NAMING_RULES) {
    if (r.check(parsed, dirName, merged)) {
      issues.push(makeIssue(r.code, r.severity, r.field, r.message));
    }
  }
  return issues;
}
